src/index.js

```diff
@@ -0,0 +1,194 @@
+          'the parent is always responsible for deciding whether or not to render any of its children.',
+          'It is not the child\'s responsibility to decide whether a render instruction from above should',
+          'get ignored or not by returning `null`.\n\nWhen any component gets its render() function called,',
+          'that is the signal that it should be rendering its part of the UI. It may in turn decide not to',
+          'render all of *its* children, but it should never return `null` for itself. It is not responsible',
+          'for that decision.'
+        ].join(' '));
+      }
+
+      const fn = this.__outsideClickHandler = generateOutsideCheck(
+        componentNode,
+        instance,
+        clickOutsideHandler,
+        this.props.outsideClickIgnoreClass,
+        this.props.excludeScrollbar,
+        this.props.preventDefault,
+        this.props.stopPropagation
+      );
+
+      const pos = registeredComponents.length;
+      registeredComponents.push(this);
+      handlers[pos] = fn;
+
+      // If there is a truthy disableOnClickOutside property for this
+      // component, don't immediately start listening for outside events.
+      if (!this.props.disableOnClickOutside) {
+        this.enableOnClickOutside();
+      }
+    }
+
+    /**
+    * Track for disableOnClickOutside props changes and enable/disable click outside
+    */
+    componentWillReceiveProps(nextProps) {
+      if (this.props.disableOnClickOutside && !nextProps.disableOnClickOutside) {
+        this.enableOnClickOutside();
+      } else if (!this.props.disableOnClickOutside && nextProps.disableOnClickOutside) {
+        this.disableOnClickOutside();
+      }
+    }
+
+    /**
+     * Remove the document's event listeners
+     */
+    componentWillUnmount() {
+      this.disableOnClickOutside();
+      this.__outsideClickHandler = false;
+      const pos = registeredComponents.indexOf(this);
+      if (pos > -1) {
+        // clean up so we don't leak memory
+        if (handlers[pos]) { handlers.splice(pos, 1); }
+        registeredComponents.splice(pos, 1);
+      }
+    }
+
+    /**
+     * Can be called to explicitly enable event listening
+     * for clicks and touches outside of this element.
+     */
+    enableOnClickOutside = () => {
+      const fn = this.__outsideClickHandler;
+      if (typeof document !== 'undefined') {
+        let events = this.props.eventTypes;
+        if (!events.forEach) {
+          events = [events];
+        }
+        events.forEach(eventName => document.addEventListener(eventName, fn));
+      }
+    }
+
+    /**
+     * Can be called to explicitly disable event listening
+     * for clicks and touches outside of this element.
+     */
+    disableOnClickOutside = () => {
+      const fn = this.__outsideClickHandler;
+      if (typeof document !== 'undefined') {
+        let events = this.props.eventTypes;
```

Comment:
We probably need to do some additional tracking for eventType un/listening in componentWillReceiveProps too, given what the code's doing here, so that if the eventTypes change, we unlisten for removed types, and start listening for new types.